drivers/clock_control/clock_control_nrf.c

```diff
@@ -441,16 +440,64 @@ static void lfclk_spinwait(nrf_clock_lfclk_t t)
 {
 	nrf_clock_domain_t d = NRF_CLOCK_DOMAIN_LFCLK;
 	nrf_clock_lfclk_t type;
+	static const bool lfxo_target =
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_XTAL) ||
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_LOW_SWING) ||
+		    IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING);
+	bool isr_mode = k_is_in_isr() || k_is_pre_kernel();
+	int key = isr_mode ? irq_lock() : 0;
+
+	if (!isr_mode) {
+		nrf_clock_int_disable(NRF_CLOCK, NRF_CLOCK_INT_LF_STARTED_MASK);
+	}
 
-	while (!(nrfx_clock_is_running(d, (void *)&type) && (type == t))) {
-		/* empty */
+	while (!(nrfx_clock_is_running(d, (void *)&type)
+		 && ((type == t) || (type == CLOCK_CONTROL_NRF_K32SRC)))) {
```

Comment:
This won't work for `CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_LOW_SWING` and `CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING`, as `type` will be `NRF_CLOCK_LFCLK_Xtal` then.

I'd propose to correct it and simplify a bit the logic used here as follows.
First, to change the function signature to:
```
static void lfclk_spinwait(enum nrf_lfclk_start_mode start_mode)
```
then instead of `lfxo_target` use something like:
```
	nrf_clock_lfclk_t target_type =
		/* For sources XTAL, EXT_LOW_SWING, and EXT_FULL_SWING,
		 * NRF_CLOCK_LFCLK_Xtal is returned as the type of running clock.
		 */
		(IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_XTAL) ||
		 IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_LOW_SWING) ||
		 IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING))
		? NRF_CLOCK_LFCLK_Xtal
		: CLOCK_CONTROL_NRF_K32SRC;
 ```
and then use here:
```
	while (!(nrfx_clock_is_running(d, (void *)&type)
		 && ((type == target_type)
		     || (mode == NRF_LFCLK_START_MODE_SPINWAIT_AVAILABLE))) {
```
and below:
```
		if (target_type == NRF_CLOCK_LFCLK_Xtal
		    && (nrf_clock_lf_src_get(NRF_CLOCK) == NRF_CLOCK_LFCLK_RC)
		    && nrf_clock_event_check(NRF_CLOCK,
		    ...
```
In the `switch` in `z_nrf_clock_control_lf_on()` we'll have then:
```
	case NRF_LFCLK_START_MODE_SPINWAIT_STABLE:
	case NRF_LFCLK_START_MODE_SPINWAIT_AVAILABLE:
		lfclk_spinwait(mode);
```

Reply:
Nice suggestion. Done.